jumeg_utils.py

```diff
@@ -173,39 +234,186 @@ def make_surrogates_sklearn(epochs):
+    if (check_power):
+        ps1 = np.abs(np.fft.fft(surr))**2
+        ps2 = np.abs(np.fft.fft(epochs.get_data()))**2
+        assert ps1.all() == ps2.all(), 'The power content does not match. Error.'
+
     return surrogate
     
-def make_surrogates_shuffling(epochs):
-    ''' 
-    Make surrogate epochs by simply shuffling. Destroy time-phase relationship for each trial.
+# def make_surrogates_epoch_numpy(epochs):
+#     ''' 
+#     Make surrogate epochs by simply shuffling. Destroy time-phase relationship for each trial.
     
+#     Parameters
+#     ----------
+#     Epochs Object.
+
+#     Output
+#     ------
+#     Surrogate Epochs object
+#     '''
+#     surrogate = epochs.copy()
+#     surr = surrogate.get_data()
+#     for trial in range(len(epochs)):
+#         for channel in range(len(epochs.ch_names)):
+#             np.random.shuffle(surr[trial, channel, :])
+#     surrogate._data = surr
+#     ps1 = np.abs(np.fft.fft(surr))**2
+#     ps2 = np.abs(np.fft.fft(epochs.get_data()))**2
+#     assert ps1.all() == ps2.all(), 'The power content does not match. Error.'
+#     return surrogate
+
+
+#######################################################
+#                                                     
+# make surrogates CTPS phase trials
+#                                                     
+#######################################################
+def make_surrogates_ctps(phase_array, nrepeat=1000, mode='shuffle', n_jobs=4, 
+                        verbose=None):
+    ''' calculate surrogates from an array of (phase) trials 
+        by means of shuffling the phase
+
     Parameters
     ----------
-    Epochs Object.
+    phase_trial : 4d ndarray of dimension [nfreqs x ntrials x nchan x nsamples]
+
+    Optional:
+    nrepeat: 
+
+    mode: 2 different modi are allowed.
+        'mode=shuffle' whill randomly shuffle the phase values. This is the default
+        'mode=shift' whill randomly shift the phase values
+    n_jobs: number of cpu nodes to use
+    verbose:  verbose level (does not work yet)
+    fnout: 'fnout=filename' will save the surrogates to (numpy) file
+
+
+    Returns
+    -------
+    pt : shuffled phase trials
 
-    Output
-    ------
-    Surrogate Epochs object
     '''
-    surrogate = epochs.copy()
-    surr = surrogate.get_data()
-    for trial in range(len(epochs)):
-        for channel in range(len(epochs.ch_names)):
-            np.random.shuffle(surr[trial, channel, :])
-    surrogate._data = surr
-    ps1 = np.abs(np.fft.fft(surr))**2
-    ps2 = np.abs(np.fft.fft(epochs.get_data()))**2
-    assert ps1.all() == ps2.all(), 'The power content does not match. Error.'
-    return surrogate
+
+    from joblib import Parallel, delayed
+    from mne.parallel import parallel_func
+    from mne.preprocessing.ctps_ import kuiper
+    
+    nfreq, ntrials, nsources, nsamples  = phase_array.shape
+    ks = np.zeros((nfreq,nrepeat,nsources, nsamples))        # Kuiper's stat
```

Comment:
Yes, this is in general  a problem of computer hardware.
I am running a 1000 repetitions using 611 samples, 117 trials, 4 freq bands and for the test 15-20 components. 
Running 100 repetitions only is too few. Actually only need the significance value and not the Kuiper's stat value; we could save 50% of the memory needed.